idb/idb.go

```diff
@@ -173,7 +173,7 @@ type IndexerDb interface {
 	// The next multiple functions return a channel with results as well as the latest round
 	// accounted.
 	Transactions(ctx context.Context, tf TransactionFilter) (<-chan TxnRow, uint64)
-	GetAccounts(ctx context.Context, opts AccountQueryOptions) (<-chan AccountRow, uint64)
+	GetAccounts(ctx context.Context, opts AccountQueryOptions) (<-chan AccountRow, uint64, *bookkeeping.BlockHeader)
```

Comment:
@tolikzinovyev I'm curious what you think about this. The protocol is needed for the min balance calculation, but this approach puts accounting logic in the REST API. There are a few other bits of accounting logic outside of the eval function (sig type, created/deleted at, pending rewards calculations, maybe others). Is there a way to make this sort of accounting logic pluggable so that it can be encapsulated so that it doesn't need to be part of the DB/REST API?

cc: @AlgoStephenAkiki @shiqizng

Reply:
I think the right solution is to return `basics.AccountData`s in `IndexerDb.GetAccounts()` like it's done in other `IndexerDb` functions, and in `api` invoke `basics.AccountData.MinBalance()`.

Reply:
That seems reasonable, but it still has the same problem, you need ConsensusParams to pass into MinBalance.

Reply:
I wouldn't say it's a problem. `GetAccounts()` returns the current round. Similarly, it can return the current consensus version.

Reply:
👍

Reply:
Look further into this issue I have some thoughts regarding @tolikzinovyev suggestions.

1. modifying `GetAccounts()` to return the current consensus version is sensible and straightforward
2. but having `GetAccounts()` return a `basics.AccountData` is more problematic. Let me explain why and get your thoughts.

We would need to modify the `<-chan AccountRow` output to return a `basics.AccountData`. We have two options for how to do this:
#### A) return it while also keeping the `models.Account` that is currently being returned
#### B) replace the `models.Account` that is being returned by this new `basics.AccountData`

Option **A** has the _advantage_ of making everything backwards compatible downstream. We only need to have the new functionality, such as `MinBalance`, opt in to this structure. It also has the _disadvantage_ of creating 2 parallel code paths for dealing with accounts. It will be difficult to know which of the two structures has the "real" account information.

Option **B** has the _advantage_ of providing the canonical account information. It has the  _disadvantage_ that current downstream consumers of `models.Account` will need to somehow consume a `basics.AccountData` instead. So we would need to either re-write the these consumers, or provide a solid adaptors that convert the models as needed. There are already similar adaptors in the `go-algorand` repo, but there are some subtle differences between `indexer` and `go-algorand` so I will not be able to use those adaptors "out of the box". Thus, we would likely also end up with parallel adaptors in each of the repos. In particular, here are the current consumers of the `models.Account` that `GetAccounts()` is currently providing:
* _account rewind functionality_: This seems to be basically broken as it cannot handle some relatively new transactions. Putting effort into writing an adaptor for this function seems to be misguided in this PR given the state of the functionality.
* endpoint `GET /v2/accounts/{account-id}`: would need to convert the models - but this seems like a reasonable goal, if we decide to embark down this path
* endpoint `GET /v2/accounts`: similar to the previous, we would need to convert the models, but it seems like a worthwhile endeavor

Reply:
## What am I proposing to do?

Because I believe any effort to upgrade `rewind()` is misguided in this PR, it suggests the "uglier" parallel model approach. But now that we are providing both models, it means that keeping the PR focused only on minbalance is simplest. I.e., I propose to have `GetAccounts()` output of `<-chan AccountRow` have its type modified as follows:
```go
type AccountRow struct {
	Account models.Account
        MinBalanceAccount basics.AccountData
	Error   error
}
```
the second field `MinBalanceAccount` is new. As the name suggests, the purpose of this field is to enable the calculation of `MinBalance()` so it would not be a "full" account, and the source of account truth would still be the `models.Account` field. I also propose to create new issues about ameliorating this non-ideal situation. The issues I would create would be:
1. Add adaptors to `indexer` which convert between `basics.AccountData` and `models.Account` (these would be similar to the ones already in `go-algorand`)
2. When `rewind()` gets fixed, it should use `basics.AccountData` and at the same time, `MinBalanceAccount` in `AccountRow` should be upgraded to the `Account` and have all the information that was in the old `models.Account`. The other downstream functionality would get their `models.Account` translated from a `basics.AccountData` right before returning their responses to the client.

Reply:
### An Even Quicker Approach

If we're aiming to get this PR out as soon as possible, an even quicker option is **C** which doesn't return any `basics.AccountData`. I could still push the calc of `MinBalance` closer to the view without a large effort. In this case, I would still create issues similar to the above.

Reply:
I think we need to wait for the ledger refactoring changes. Whether or not these approaches will continue working depend on changes that are yet to be made. For example, do we add the asset counts lazily or as a migration? If it's done lazily we may not always be able to compute the min balance depending on the REST API changes for the ledger refactoring (one proposal includes an option to return account data with asset/application data removed).

Regardless of that, I think we should skip the `basics.AccountData` conversion. Does it really help solve our problem? We're constructing it from metadata outside of algod, so we may end up with data inconsistencies based on how we reassemble it from Postgres.

Reply:
Agree with the first part, but ideally the database layer shouldn't construct `api` types imo. Constructing `basics.AccountData` from data in postgres is pretty straightforward, it's already done in `idb/postgres/internal/ledger_for_evaluator/`.

Reply:
I support Will's position that we wait for the ledger refactoring changes. We don't want to add this now, and then have updates to this work become a blocker for the ledger refactoring project (we have enough complex dependencies there already). 
People can continue to work without min balance being easily available for a little bit longer.

Reply:
![image](https://user-images.githubusercontent.com/291133/151870705-19911bab-8088-4a70-8323-7b4c5df8a377.png)